File: apps/web/lib/wake.ts
/**
 * Waking a PC from the dashboard: which of the owner's PCs can send the packet, and what to say about it.
 *
 * A sleeping PC cannot hear a command, so the wake is sent to another PC that is online and broadcasts a
 * magic packet on its own local networks. WOLF cannot tell which PCs share a network — it never learns
 * anybody's LAN layout — so the owner chooses the sender, and the page says that it has to be on the same
 * network as the one asleep.
 */

/** The parts of a PC this needs. Structural, so the dashboard's own PC type fits. */
export interface WakeablePc {
  readonly id: string;
  readonly name: string;
  readonly status: string;
  readonly remoteAccessEnabled: boolean;
  readonly capabilities: {
    readonly wakeOnLanCapable: boolean;
    readonly wakeAddressKnown?: boolean;
    readonly supportedCommands: readonly string[];
  } | null;
}

export type WakeReadiness =
  | { readonly kind: 'online' }
  | { readonly kind: 'switched-off' }
  | { readonly kind: 'no-address' }
  /** An address is known, but Windows on that PC had not armed the adapter the last time it connected. */
  | { readonly kind: 'not-armed' }
  | { readonly kind: 'ready' };

export function wakeReadiness(target: WakeablePc): WakeReadiness {
  if (target.status === 'online') return { kind: 'online' };
  if (!target.remoteAccessEnabled) return { kind: 'switched-off' };
  if (!target.capabilities?.wakeAddressKnown) return { kind: 'no-address' };
  if (!target.capabilities.wakeOnLanCapable) return { kind: 'not-armed' };
  return { kind: 'ready' };
}

/** The owner's other PCs that can send a wake packet right now, by name. */
export function wakeSenders<T extends WakeablePc>(target: WakeablePc, pcs: readonly T[]): T[] {
  return pcs
    .filter((pc) => pc.id !== target.id)
    .filter((pc) => pc.status === 'online' && pc.remoteAccessEnabled)
    .filter((pc) => pc.capabilities?.supportedCommands.includes('power.wake') === true)
    .sort((a, b) => a.name.localeCompare(b.name));
}

export interface WakeResult {
  readonly packetsSent: number;
  readonly networks: number;
}

/** What was done, and only that: a broadcast cannot know whether anything woke. */
export function wakeSentText(sender: string, target: string, result: WakeResult): string {
  const packets = `${result.packetsSent} wake packet${result.packetsSent === 1 ? '' : 's'}`;
  const networks = `${result.networks} local network${result.networks === 1 ? '' : 's'}`;
  return (
    `${sender} sent ${packets} for ${target} across ${networks}. ` +
    `WOLF shows ${target} online when its agent connects, usually within a minute of it waking. ` +
    `If it does not, it may be on a different network from ${sender}, or its firmware may not allow waking.`
  );
}

File: apps/android/app/src/main/java/app/amizhthan/wolf/api/Wake.kt
package app.amizhthan.wolf.api

import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonObject

/**
 * Waking a PC from the phone — the web dashboard's rules, word for word.
 *
 * A sleeping PC cannot hear a command, so the wake goes to another of the owner's PCs that is online, which
 * broadcasts a magic packet on its own local networks. WOLF never learns which PCs share a network, so the
 * owner chooses the sender. The phone sends only which PC to wake; the address is filled in by the API from
 * what that PC reported, and the phone never sees it.
 */
object Wake {
    enum class Readiness { ONLINE, SWITCHED_OFF, NO_ADDRESS, NOT_ARMED, READY }

    fun command(targetPcId: String): JsonObject = buildJsonObject {
        put("type", "power.wake")
        putJsonObject("payload") { put("targetPcId", targetPcId) }
    }

    fun readiness(target: PcSummary): Readiness = when {
        target.status == "online" -> Readiness.ONLINE
        !target.remoteAccessEnabled -> Readiness.SWITCHED_OFF
        target.capabilities?.wakeAddressKnown != true -> Readiness.NO_ADDRESS
        !target.capabilities.wakeOnLanCapable -> Readiness.NOT_ARMED
        else -> Readiness.READY
    }

    /** The owner's other PCs that can send a wake packet now, by name. */
    fun senders(target: PcSummary, pcs: List<PcSummary>): List<PcSummary> = pcs
        .filter { it.id != target.id && it.status == "online" && it.remoteAccessEnabled }
        .filter { it.capabilities?.supportedCommands?.contains("power.wake") == true }
        .sortedBy { it.name.lowercase() }

    /** What was done, and only that: a broadcast cannot know whether anything woke. */
    fun sentText(sender: String, target: String, packetsSent: Int, networks: Int): String =
        "$sender sent $packetsSent wake packet${if (packetsSent == 1) "" else "s"} for $target across $networks local " +
            "network${if (networks == 1) "" else "s"}. WOLF shows $target online when its agent connects, usually within a minute " +
            "of it waking. If it does not, it may be on a different network from $sender, or its firmware may not allow waking."
}
